### ISSUE TITLE
R1-E3M: label-free real-track ESN memory validation

### ISSUE THREAD

**Issue by qigao** [qigao/yolo-motion-perception#15]
## Objective

Validate the **mechanism-level ESN claim** on immutable real YOLO/BoT-SORT
trajectories before semantic behavior annotation is complete:

> Does a fixed reservoir state retain linearly decodable information about
> earlier real-track observations that is absent from the instantaneous current
> observation, and does that advantage collapse when recurrent history is
> deliberately destroyed?

This issue is intentionally **label-free with respect to human behavior
semantics**. It does not classify `approach/touch/pick_up/pass_by` and does not
replace #11 or the registered semantic R1-E3 experiment in #8 / PR #9.

## Why this is separate

The current semantic R1-E3 design asks whether temporal reservoir
representations improve four-class real behavior classification. That claim
requires independent human event labels.

This issue asks an earlier and narrower question: whether ESN history itself is
useful on real detector/tracker trajectories. Ground truth is generated from
the trajectory's own earlier observations, not from human semantic judgment.

Therefore:

- #11 remains mandatory for semantic claims;
- #11 is **not** a dependency for this mechanism experiment;
- no semantic labels may be inferred or substituted by YOLO/KPF/model output.

## Frozen inputs

Use only source videos already frozen by #10/#11:

- Batch-1: 11 MEVA videos;
- 7 train / 4 eval;
- 4 train / 2 eval guarded source-window components;
- source manifest SHA-256:
  `1db8cc200ba9eaab2c14ed260b06ed582ed75d54f57c3036e1312cd2f184f4cf`;
- source-video artifact digest:
  `sha256:211e41cf0d6100b13b8194dc81d2faf7dcb492984459de674f0eefe5fcacc838`.

The existing extraction contract from #8 is reused. Detector/tracker settings
must be frozen before any mechanism score is computed.

## Artifact boundary

```text
frozen real videos
      ↓
fixed YOLO11 + fixed BoT-SORT
      ↓
immutable raw per-track rows
      ↓
deterministic track-window construction
================ frozen artifact boundary ================
fixed ESN reservoirs
      ↓
linear delayed-observation probes
      ↓
history-destruction controls
```

No human semantic annotation appears in this pipeline.

## Deterministic track-window construction

Build windows only from one track identity within one source video.

For each eligible track:

- use normalized image-space channels:
  `cx, cy, w, h, detector_confidence, presence`;
- never interpret bbox size as metric depth;
- resample deterministically to 20 temporal bins;
- reject a window if interpolation/gap rules cannot produce finite data;
- never bridge a tracker-ID discontinuity;
- never cross source-video boundaries;
- keep the already-frozen whole-video train/eval split;
- use deterministic non-overlapping or predeclared-stride windows only;
- freeze all window IDs/digests before ESN scoring.

Track IDs are provenance only and are not model features.

## Frozen ESN configuration

Inherit without search from R1-E2 / R1-E3:

- architectures: Flat-256, Grouped-4x64, Hierarchical-2x128,
  Hierarchical-4x64;
- total neurons: 256;
- tanh;
- spectral radius: 0.9;
- leak: 1.0;
- Ridge regularization: `1e-6`;
- seeds: `[7, 17, 29, 43, 61]`;
- float64;
- no learned recurrent weights;
- no topology/window/seed/hyperparameter search.

Architecture is a blocking factor, not a selection target.

## Primary probe: delayed observation reconstruction

At normalized bin 19, fit a linear Ridge probe to reconstruct earlier
track-observation geometry from:

- **M0 — instantaneous:** current 6-channel observation at bin 19;
- **M1 — reservoir:** current reservoir state after bins 0..19.

Freeze delays before measurement:

`[1, 2, 5, 10, 15]` bins.

For each delay, the target is the earlier geometric observation
`[cx, cy, w, h]`. Detector confidence/presence may be input channels but are
not reconstruction targets.

Fit probes on training videos only and evaluate on frozen evaluation videos.

Primary per-delay metric:

- mean evaluation R² over the four target channels.

Also retain:

- per-channel R²;
- MSE;
- prediction/coefficient digests;
- valid sample count.

Primary contrast:

```text
delta_R2(delay) = R2(M1 reservoir) - R2(M0 instantaneous)
```

This directly measures whether the reservoir exposes past trajectory information
that the current frame alone does not.

## History-destruction controls

Use the exact fitted M1 probe; never refit controls.

### H1 — recurrent reset

For each evaluation window:

1. process bins 0..15;
2. reset reservoir immediately before bin 16;
3. process bins 16..19;
4. evaluate the same delayed-reconstruction probe.

This preserves the final four observations while removing earlier recurrent
state.

### H2 — prefix permutation

For each evaluation window:

- keep bins 16..19 unchanged;
- deterministically permute bins 0..15 using a frozen seed derived from the
  window ID;
- evaluate with the same fitted probe.

No control may change the target value or refit the readout.

## Secondary mechanism diagnostic: same-suffix / different-history pairs

Before scoring, deterministically construct audit pairs whose final four input
bins are close in input space while their earlier prefixes are materially
different.

For each pair report:

- suffix input distance;
- prefix input distance;
- final reservoir-state distance;
- reset-control reservoir-state distance.

This is diagnostic only and cannot replace the delayed-reconstruction primary
probe.

## Leakage controls

- train/eval isolation remains whole-source-video;
- no window from an eval video enters Ridge fitting;
- no human behavior label is read;
- no MEVA KPF activity label is used as a target;
- no semantic event boundary is used;
- no detector/tracker/model score may alter split membership after freeze;
- windows/pairs/delays are frozen before ESN results are computed;
- controls reuse fitted probes and cannot refit.

## Interpretation boundary

A positive result may support only:

> Fixed ESN reservoir states retain linearly decodable information about earlier
> real YOLO/BoT-SORT track observations beyond the instantaneous current
> observation, under the frozen real-track corpus and controls.

It does **not** establish:

- behavior understanding;
- approach/touch/pick_up/pass_by recognition;
- semantic action recognition;
- physical velocity or metric depth;
- superiority of one reservoir topology;
- detector/tracker accuracy.

Those remain separate claims.

## Acceptance gates

- [ ] frozen extraction manifest bound to the existing 11 source videos;
- [ ] immutable raw-track artifact with exact digest;
- [ ] deterministic track-window artifact frozen before ESN scoring;
- [ ] zero source-video overlap across train/eval;
- [ ] delays exactly `[1,2,5,10,15]`;
- [ ] 4 architectures x 5 seeds = exactly 20 arms;
- [ ] M0/M1 use the same frozen train/eval windows;
- [ ] H1/H2 use the already-fitted M1 probe with no refit;
- [ ] report per-delay and aggregate deltas without topology selection;
- [ ] freeze raw evidence and exact-head/runtime provenance;
- [ ] semantic #11 labels are never read by this experiment.

## Relationship to existing work

- #8 / PR #9 remain the **semantic transfer** experiment.
- #10/#11 provide the already-frozen real source-video set.
- #11 human annotation remains required for semantic R1-E3 but is not required
  here.
- This issue can proceed immediately through extraction/window freeze while #11
  human review continues independently.

**Comment by qigao** [comment]
R1-E3M implementation checkpoint is now exact-head green through Tasks 1–7.

- exact head: `93f2755fd943549955fb76fc6ef5f56bcb0a857b`
- focused workflow: `35443937343` — success
- focused tests: success
- Ruff: success
- semantic-coupling fail-closed gate: success
- no-extraction/no-registered-measurement workflow gate: success

Implemented:
1. label-free mechanism artifact schema;
2. strictly causal 2.0 s / 20-bin per-track windows;
3. candidate/frozen artifact packaging and verification;
4. immutable train/eval mechanism dataset loader;
5. delayed reconstruction probes at [1,2,5,10,15];
6. no-refit reset/permutation history controls;
7. exact 20-arm memory benchmark and M-A/B/C classifier.

Protocol hardening during implementation: window resampling was changed before measurement from future-aware interpolation to strictly causal latest-observation hold with max age 0.5 s. A future detector/tracker observation can never populate an earlier bin.

No real extraction, real ESN score, prospective seal, or registered measurement has run.

**Comment by qigao** [comment]
R1-E3M code/protocol checkpoint is now closed through Tasks 1–10.

- exact head: `059aa0bdc6a4b48748a0d89bb5ad4ddf305123a9`
- focused mechanism tooling run: `35447010825` — **success**
- focused tests: success
- Ruff: success
- semantic-coupling fail-closed gate: success
- ordinary CI extraction/measurement prohibition: success

Protocol hardening now additionally freezes:

- causal source-time windows (no future observation fill);
- target-bin presence eligibility rather than treating missing geometry as zero truth;
- one shared per-delay sample mask across M0/M1/H1/H2;
- >=20 train / >=10 eval target-valid windows per delay;
- >=2 source videos per split per delay;
- content-addressed per-delay window-ID registration;
- content-addressed same-suffix/different-history pair set;
- prospective binding of artifact root + pair set + delay registration + exact scientific head/runtime.

No accepted Task 11 extraction artifact exists yet, no prospective measurement
seal has been approved, and no registered ESN mechanism score has run.

**Comment by qigao** [comment]
R1-E3M Task 11 sealed extraction has been explicitly triggered.

- trigger / exact extraction head: `665203d9b9d3a02763459cd0614265a607d41873`
- workflow: `R1-E3M sealed real-track extraction`
- frozen input: MEVA Batch-1 artifact from run `35434084445`
- source manifest SHA-256: `1db8cc200ba9eaab2c14ed260b06ed582ed75d54f57c3036e1312cd2f184f4cf`

The workflow is required to pass:
1. extraction/tooling preflight;
2. exact 11-video byte/hash verification;
3. sealed YOLO11 + BoT-SORT pass 1;
4. artifact freeze/verify;
5. independent replay pass 2;
6. byte-identical raw-track, provenance, candidate tensor, and frozen-manifest comparison;
7. accepted artifact upload.

No semantic labels are read and no ESN mechanism score is computed.

Task 12 prospective sealing will not be prepared unless this Task 11 workflow produces an accepted artifact root digest.

**Comment by qigao** [comment]
Explicit human approval received for the single R1-E3M registered measurement.

Approved exact registration:

- scientific head: `69f319089dd95296fc0e6ae29194017f8c6f874b`
- artifact root digest: `3bd81e1e97d1996fd2447754ff3c654f2dfb69965a949dbe0b18ddd855ba5ed4`
- prospective manifest SHA-256: `3202074b48852f609595426f549f4de190ed32e502123062938ae24758b2c817`
- runtime: Python 3.12.14 / NumPy 2.5.3
- arms: 20
- delays: [1,2,5,10,15]
- history-pair digest: `e7de29c58c6626d3a31f01653b7913eabd7e001b6025328b275d5219dbf26ab8`
- delay-registration digest: `6d89fa54881d3677f0402077f0e3f02fea6c3473c85d50d601ba11e3de4f1336`

Authorization: execute exactly one registered measurement against this exact sealed manifest. No refit, resealing, parameter change, artifact substitution, or second measurement is authorized.

**Comment by qigao** [comment]
R1-E3M lifecycle audit correction: a concurrent workflow had already consumed the approved registered-measurement slot before run `35476129790` was started.

Canonical registered result remains the **first authorized run**:

- registered workflow run: `35474844878`
- scientific head: `69f319089dd95296fc0e6ae29194017f8c6f874b`
- manifest SHA-256: `3202074b48852f609595426f549f4de190ed32e502123062938ae24758b2c817`
- frozen result SHA-256: `af8f05822f614b2367368161ad68708bcca6137562e5d42dc83f4822227a1d85`
- permanent evidence commit: `93844b5659278c2d23feae7bd4bc876ed37f9d2b`
- frozen path: `docs/experiments/r1-e3m-real-track-memory`
- registered outcome: `M-A`

A second run, `35476129790`, was accidentally executed because a parallel agent/workflow had already claimed and completed the slot, while the later guard checked only for its own guard artifact and did not detect the earlier persistent issue-slot claim. This is a protocol incident.

The second run is **excluded from scientific interpretation and is not a replacement or replicate**. Its numerical summary is nearly identical to the canonical run, but it must not be used to strengthen the registered claim.

Corrective actions:
- the reintroduced one-shot measurement workflow was deleted in commit `e86e48f9a6d36b09beec0920d0366a9f5ba52e71`;
- the reintroduced measurement trigger was deleted in commit `2cafbb9413f7a223fdc813159e5351f29980194e`;
- no further registered measurement is authorized;
- the in-repository frozen evidence from run `35474844878` remains the sole canonical registered evidence.

This comment supersedes any implication that run `35476129790` was the unique registered run.

**Comment by qigao** [comment]
R1-E3M final cleanup gate is green.

Canonical registered evidence remains the first authorized/frozen run:

- scientific head: `69f319089dd95296fc0e6ae29194017f8c6f874b`
- canonical registered run: `35474844878`
- frozen result SHA-256: `af8f05822f614b2367368161ad68708bcca6137562e5d42dc83f4822227a1d85`
- frozen evidence commit: `93844b5659278c2d23feae7bd4bc876ed37f9d2b`
- outcome: `M-A`

Cleanup head:

- head: `2cafbb9413f7a223fdc813159e5351f29980194e`
- E3M focused CI: `35476502542` — success
- full CI: `35476502557` — success
- Phase C4-A preflight: `35476502535` — success
- reintroduced one-shot workflow and trigger: removed
- no second registered measurement is authorized

Protocol incident note remains in this issue: accidental later run `35476129790` is excluded from scientific interpretation and does not replace or strengthen the canonical registered result.

R1-E3M Tasks 1–13 are therefore complete from a scientific/evidence lifecycle perspective. PR #16 remains Draft and is not merged.

**Comment by qigao** [comment]
Canonical merge candidate is now clean standalone PR #17.

Final merge candidate:
- head: `431308a38cc3a46faab04fed6ddf1c53be936c6d`
- base: merged R1-E2 `84eb8775b8eecc65cf14a8c40cf0d7a6370bd364`
- commits: 1
- ahead/behind: 1 / 0
- mergeable: true
- E3M focused CI `35494779760`: success
- full CI `35494779752`: success
- C4-A preflight `35494779755`: success

PR #17 contains no semantic R1-E3 ancestry, no extraction/prospective regeneration
workflow, and no registered-measurement trigger. Permanent CI is verification-only.

The canonical frozen result and scientific interpretation are unchanged.

**Comment by qigao** [comment]
Final merge record:

- clean standalone PR #17 merged successfully;
- merged head: `431308a38cc3a46faab04fed6ddf1c53be936c6d`;
- merge commit: `04d2b8db33ccae22b5a904811bab1bb69d125e31`;
- exact-head E3M mechanism tooling / full CI / C4-A preflight were all green before merge;
- canonical registered run remains `35474844878`;
- canonical result SHA-256 remains `af8f05822f614b2367368161ad68708bcca6137562e5d42dc83f4822227a1d85`;
- registered outcome remains `M-A`, 20/20 arms positive;
- accidental later run `35476129790` remains excluded from scientific interpretation.

No semantic R1-E3 claim is implied by this merge; #9/#11 remain independent.

**Comment by qigao** [comment]
Post-merge verification is fully green on merge commit `04d2b8db33ccae22b5a904811bab1bb69d125e31`.

- full CI `35503505776` — success;
- R1-E3M mechanism tooling `35503505775` — success;
- Phase C4-A preflight `35503507039` — success;
- no failed post-merge jobs.

This closes the repository-integration gate for R1-E3M. Canonical scientific evidence remains unchanged: registered run `35474844878`, result SHA-256 `af8f05822f614b2367368161ad68708bcca6137562e5d42dc83f4822227a1d85`, outcome `M-A`, 20/20 positive arms. Semantic R1-E3 remains isolated in #9/#11.
